internal/installercache/installercache_test.go

```diff
@@ -118,20 +147,173 @@ var _ = Describe("installer cache", func() {
 			Expect(l.extractDuration).ShouldNot(BeZero())
 		}
 		Expect(l.Path).ShouldNot(BeEmpty())
-
-		time.Sleep(1 * time.Second)
 		Expect(l.startTime.Before(time.Now())).To(BeTrue())
+		Expect(l.Cleanup(context.TODO())).To(Succeed())
 		return fname, l.Path
 	}
-	It("evicts the oldest file", func() {
+
+	type test struct {
+		releaseID string
+		version   string
+		clusterID strfmt.UUID
+	}
+
+	getUsedBytesForDirectory := func(directory string) uint64 {
+		var totalBytes uint64
+		seenInodes := make(map[uint64]bool)
+		err := filepath.Walk(directory, func(path string, fileInfo os.FileInfo, err error) error {
+			if err != nil {
+				if _, ok := err.(*os.PathError); ok {
+					// Something deleted the file before we could walk it
+					// count this as zero bytes
+					return nil
+				}
+				return err
+			}
+			stat, ok := fileInfo.Sys().(*syscall.Stat_t)
+			Expect(ok).To(BeTrue())
+			if !fileInfo.IsDir() && !seenInodes[stat.Ino] {
+				totalBytes += uint64(fileInfo.Size())
+				seenInodes[stat.Ino] = true
+			}
+			return nil
+		})
+		Expect(err).ToNot(HaveOccurred())
+		return totalBytes
+	}
+
+	runTest := func(t test, manager *Installers) (*Release, error) {
+		expectEventsSent()
+		mockReleaseCalls(t.releaseID, t.version)
+		return manager.Get(ctx, t.releaseID, "mirror", "pull-secret", mockRelease, t.version, t.clusterID)
+	}
+
+	cleanup := func(release *Release) {
+		if release != nil {
+			time.Sleep(2 * time.Millisecond)
+			Expect(release.Cleanup(ctx)).To(Succeed())
+		}
+	}
+
+	handleError := func(errMutex *sync.Mutex, err error, reportedError *error) {
+		if err != nil {
+			errMutex.Lock()
+			if reportedError == nil {
+				*reportedError = err
+			}
+			errMutex.Unlock()
+			return
+		}
+	}
+
+	// runParallelTest launches installercache.Get tests in parallel
+	// releases are automatically cleaned up as they are gathered
+	// returns the first error encountered or nil if no error encountered.
+	runParallelTest := func(maxCapacity int64, maxReleaseSize int64, tests []test) error {
+		log := logrus.New()
+		log.SetOutput(io.Discard)
+		var err error
+		manager, err = New(getInstallerCacheConfig(maxCapacity, maxReleaseSize), eventsHandler, diskStatsHelper, log)
+		Expect(err).ToNot(HaveOccurred())
+		var wg sync.WaitGroup
+		var reportedError error
+		var errMutex sync.Mutex
+		for _, t := range tests {
+			wg.Add(1)
+			go func(t test) {
+				defer wg.Done()
+				release, err := runTest(t, manager)
+				handleError(&errMutex, err, &reportedError)
```

Comment:
IIUC this would only report the first error. Writing the error on a channel would be more idiomatic and might help to both simplify and report errors correctly.

Reply:
I actually used a channel based approach before but did not like it for a number of reasons.

- It did not happen often but I found I would get the occasional test flake where all of the errors had not been collected (it was rare but I did not like this)
- I only need the first error in any case and it does not add value to have them all.
- The Mutex does a great job of making sure we can examine the error once the async jobs are done.